File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.java
class Solution {
    private boolean f(int[] weights, int days, int maxCapacity){
        int currWeight = 0, noOfDays = 0;
        for(int i : weights){
            currWeight += i;
            if(currWeight >= maxCapacity){
                noOfDays++;
                currWeight = (currWeight == maxCapacity) ? 0 : i;
            }
        }
        noOfDays += (currWeight != 0 ? 1 : 0);
        return noOfDays <= days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int low = Arrays.stream(weights).max().getAsInt();
        int high = Arrays.stream(weights).sum(), ans = high;
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(f(weights, days, mid)){
                ans = mid;
                high = mid - 1;
            } else{
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
private:
    int f(vector<int>& weights, int days, int cap){
        int curr = 0;
        for(int i : weights){
            if(curr + i == cap){
                curr = 0; days--;
            } else if(curr + i > cap){
                curr = i; days--;
            } else{
                curr += i;
            }
        }
        days -= (curr != 0);

        return days >= 0;
    }

public:
    int shipWithinDays(vector<int>& weights, int days) {
        int l = 0, r = 0;
        for(int i : weights){
            l = max(l, i);
            r += i;
        }

        int ans = -1;
        while(l <= r){
            int m = l + (r - l) / 2;
            if(f(weights, days, m)){
                ans = m;
                r = m - 1;
            } else{
                l = m + 1;
            }
        }

        return ans;
    }
};